For #6542 - Open WindowRequest url in current tab if multitab is disabled

Mugurell opened the pull request:
The AC flow for the `window.open` Gecko api is:
- UpdateWindowRequestAction - which adds the WindowRequest to the current tab
- AddTabAction which will open a new tab - first with "about:blank" then in
which the requested window url will be loaded.
AddTabAction for "about:blank" is currently dropped if not supporting multitab.
So in scenario from a users' standpoint the app won't do anything.
- ConsumeWindowRequestAction which will cleanup the WindowRequest from the
current tab's state after the app acted upon that request.
Knowing that this flow is finished Focus will here load the url of the
requested window in the current tab while a bit sooner opening a new tab was
avoided.


https://user-images.githubusercontent.com/11428869/157041898-4b387b3a-d992-450d-bb1a-d9f9564a9f98.mp4



### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on `Show All Checks`,
2. click `Details` next to `build-focus-debug` or `build-klar-debug` for changes targeting Klar,
2. click `View task in Taskcluster`,
3. click the `Artifacts` row,
4. click to download any of the apks listed here which use an appropriate name for each CPU architecture.

Mugurell commented on app/src/main/java/org/mozilla/focus/tabs/MergeTabsMiddleware.kt:
```diff
@@ -31,6 +33,25 @@ class MergeTabsMiddleware(
     ) {
+        if (!hasTabs && action is ConsumeWindowRequestAction) {
+            val isRequestToCloseWindow =
+                context.state.privateTabs.first().content.windowRequest?.type == WindowRequest.Type.CLOSE
```
Tested on https://jsfiddle.net/da5z3egp/ and saw that while on desktop the window will be closed there is no such callback coming from GV for us.
Not sure if/how this request would get to use but the app should be ready to handle it.
In the above video I used a delayed `this.context.components.tabsUseCases.removeAllTabs()` to see how the app would act in the event such a `CLOSE` request would come to the app after a `OPEN` request was acted upon.

csadilek replied:
Here's an alternative that should work fine: https://jsfiddle.net/8dp4huag/1/

We should file a follow-up to investigate why the close call doesn't work in that `setTimeout`, but that's not a blocker for this.

Mugurell commented:
> @Mugurell The flow here makes sense to me, but tabs added by the `WindowFeature` will also come with a new`engineSession` that has to be used. Same as we discussed in [#6546 (comment)](https://github.com/mozilla-mobile/focus-android/issues/6546#issuecomment-1061125171) and fixed by #6599.
> 
> So we should test this change on top of it to make sure we're not leaking engine sessions and everything works correctly.

Tabs are indeed leaking in this `window.open` scenario.
If allowing to try loading "about:blank" [here](https://github.com/mozilla-mobile/focus-android/blob/66c8d2f27492f8799e0201d0ceeb167cda85e3ea/app/src/main/java/org/mozilla/focus/tabs/MergeTabsMiddleware.kt#L109) and as such pass [this check](https://github.com/mozilla-mobile/focus-android/blob/66c8d2f27492f8799e0201d0ceeb167cda85e3ea/app/src/main/java/org/mozilla/focus/tabs/MergeTabsMiddleware.kt#L49) and get to execute the newly updated code from https://github.com/mozilla-mobile/focus-android/pull/6599 then I get into an infinite loop of tabs being closed and recreated - seemingly from the AC [`consumeWindowRequest` method](https://github.com/mozilla-mobile/android-components/blob/7475b34e7a18cf77222520fa593b529fdf80f39a/components/feature/tabs/src/main/java/mozilla/components/feature/tabs/WindowFeature.kt#L60).

csadilek reviewed (commented):
Ok, I think this can be rebased and landed now :). The `AddTabAction` should be processed first and merge the tabs correctly (using the new engine session), then the `ConsumeWindowRequestAction` should trigger loading the url.

The back button will not work as this is essentially a new window. With multiple tabs, the back button would go to the parent session, but since we have to close the parent session (single tab browser) there's nowhere to go back to.

csadilek commented:
> Tabs are indeed leaking in this window.open scenario.

@Mugurell I've rebased this on top of main and pushed to https://github.com/csadilek/focus-android/commit/604f3dee0a49ddf1d4f274f4af62b19aea6d30c3, which all works fine for me! I don't see any session leaks or an infinite recursion either. Can you double-check, maybe something went wrong in your local branch?

Mugurell commented:
> > Tabs are indeed leaking in this window.open scenario.
> 
> @Mugurell I've rebased this on top of main and pushed to [csadilek@604f3de](https://github.com/csadilek/focus-android/commit/604f3dee0a49ddf1d4f274f4af62b19aea6d30c3), which all works fine for me! I don't see any session leaks or an infinite recursion either. Can you double-check, maybe something went wrong in your local branch?

Was still seeing leaked EngineSessions in `about:debugging`.
Updated now the patch to merge supporting "open window requests" in the newly updated support for merging tabs.

Mugurell commented:
Tested and all scenarios seems good now.
Landing for Maria to also get the chance to test this in preparation for uplift.

ionutbedregeanu commented:
@Mergifyio backport releases_v99.0

ionutbedregeanu commented:
@Mergifyio backport releases_v99.0

ionutbedregeanu commented:
@Mergifyio backport releases_v99.0

ionutbedregeanu commented:
@Mergifyio backport releases_v99.0

ionutbedregeanu commented:
@Mergifyio backport releases_v98.0
